gnome-initial-setup/gnome-initial-setup.c

```diff
@@ -156,6 +158,81 @@ destroy_pages_after (GisAssistant *assistant,
   }
 }
 
+static gint
+compare_pages_order (gconstpointer a,
+                     gconstpointer b,
+                     gpointer data)
+{
+  GHashTable *sort_table = data;
+  const gchar *id_a = ((PageData *) a)->page_id;
+  const gchar *id_b = ((PageData *) b)->page_id;
+  gint order_a = (gint) GPOINTER_TO_UINT (g_hash_table_lookup (sort_table, id_a));
+  gint order_b = (gint) GPOINTER_TO_UINT (g_hash_table_lookup (sort_table, id_b));
+
+  return order_a - order_b;
+}
+
+static void
+reorder_pages (GisDriver *driver)
+{
+  gint num_pages;
+  GKeyFile *key_file = gis_driver_get_vendor_conf_file (driver);
+  g_auto(GStrv) conf_pages_order = NULL;
+  g_autoptr(GHashTable) sort_table = NULL; /* (owned) (element-type utf8 uint) */
+  g_autoptr(GError) error = NULL;
+
+  if (key_file == NULL)
+    return;
+
+  conf_pages_order = g_key_file_get_string_list (key_file, "pages", "order",
+                                                 NULL, &error);
+  if (conf_pages_order == NULL) {
+    /* if we have the pages order key but there was an issue getting it */
+    if (!g_error_matches (error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_KEY_NOT_FOUND) &&
+        !g_error_matches (error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_GROUP_NOT_FOUND))
+      g_debug ("Couldn't get pages order from the conf file: %s", error->message);
+
+    return;
+  }
+
+  if (*conf_pages_order == NULL) {
+    g_debug ("No pages defined in the pages order from the conf file");
+    return;
+  }
+
+  /* assign the same sort weight to every page (this will make sure) that any
+   * page to be sorted differently will show up before pages that aren't in the
+   * custom sorting list */
+  sort_table = g_hash_table_new (g_str_hash, g_str_equal);
+
+  for (num_pages = 0; page_table[num_pages].page_id != NULL; ++num_pages) {
+    g_hash_table_insert (sort_table, (gchar *) page_table[num_pages].page_id,
+                         GUINT_TO_POINTER (PAGES_DEFAULT_SORT_WEIGHT));
+  }
+
+  /* assign the new sort weights from the conf file */
+  for (guint i = 0; conf_pages_order[i] != NULL; ++i) {
+    gchar *page_id = conf_pages_order[i];
+    if (g_hash_table_lookup (sort_table, page_id) == NULL) {
+      g_debug ("No page with id '%s' found while sorting the pages!", page_id);
+      continue;
+    }
+
+    g_hash_table_insert (sort_table, page_id, GUINT_TO_POINTER (i));
```

Comment:
What if there are more than 100 pages in the config file? I suggest that rather than define PAGES_DEFAULT_SORT_WEIGHT, you pass a `length` parameter to `g_key_file_get_string_list` and then use that as the default sort weight – it's guaranteed to be 1 higher than the index of any page in the config file.

Reply:
(If you follow this suggestion you will have to clamp the `gsize length` at `G_MAXINT` since gsize may be larger than gint.)